libyul/backends/evm/ControlFlowGraph.h

```diff
@@ -0,0 +1,193 @@
+#include <libyul/Dialect.h>
+#include <libyul/Scope.h>
+
+#include <functional>
+#include <list>
+#include <vector>
+
+namespace solidity::yul
+{
+
+/// The label pushed as return label before a function call, i.e. the label the call is supposed to return to.
+struct FunctionCallReturnLabelSlot
+{
+	std::reference_wrapper<yul::FunctionCall const> call;
+	bool operator==(FunctionCallReturnLabelSlot const& _rhs) const { return &call.get() == &_rhs.call.get(); }
+	bool operator<(FunctionCallReturnLabelSlot const& _rhs) const { return &call.get() < &_rhs.call.get(); }
+	static constexpr bool canBeFreelyGenerated = true;
+};
+/// The return label of a function while generating the code of the function body.
+struct FunctionReturnLabelSlot
+{
+	bool operator==(FunctionReturnLabelSlot const&) const { return true; }
+	bool operator<(FunctionReturnLabelSlot const&) const { return false; }
+	static constexpr bool canBeFreelyGenerated = false;
+};
+/// A slot containing the current value of a particular variable.
+struct VariableSlot
+{
+	std::reference_wrapper<Scope::Variable const> variable;
+	std::shared_ptr<DebugData const> debugData{};
+	bool operator==(VariableSlot const& _rhs) const { return &variable.get() == &_rhs.variable.get(); }
+	bool operator<(VariableSlot const& _rhs) const { return &variable.get() < &_rhs.variable.get(); }
+	static constexpr bool canBeFreelyGenerated = false;
+};
+/// A slot containing a literal value.
+struct LiteralSlot
+{
+	u256 value;
+	std::shared_ptr<DebugData const> debugData{};
+	bool operator==(LiteralSlot const& _rhs) const { return value == _rhs.value; }
+	bool operator<(LiteralSlot const& _rhs) const { return value < _rhs.value; }
+	static constexpr bool canBeFreelyGenerated = true;
+};
+/// A slot containing the idx-th return value of a previous call.
+struct TemporarySlot
+{
+	/// The call that returned this slot.
+	std::reference_wrapper<yul::FunctionCall const> call;
+	/// Specifies to which of the values returned by the call this slot refers.
+	/// index == 0 refers to the slot deepest in the stack after the call.
+	size_t index = 0;
+	bool operator==(TemporarySlot const& _rhs) const { return &call.get() == &_rhs.call.get() && index == _rhs.index; }
+	bool operator<(TemporarySlot const& _rhs) const { return std::make_pair(&call.get(), index) < std::make_pair(&_rhs.call.get(), _rhs.index); }
+	static constexpr bool canBeFreelyGenerated = false;
+};
+/// A slot containing an arbitrary value that is always eventually popped and never used.
+/// Used to maintain stack balance on control flow joins.
+struct JunkSlot
+{
+	bool operator==(JunkSlot const&) const { return true; }
+	bool operator<(JunkSlot const&) const { return false; }
+	static constexpr bool canBeFreelyGenerated = true;
+};
+using StackSlot = std::variant<FunctionCallReturnLabelSlot, FunctionReturnLabelSlot, VariableSlot, LiteralSlot, TemporarySlot, JunkSlot>;
+/// The stack top is usually the last element of the vector.
```

Comment:
Why usually?

Reply:
Well, I can (and do) use this type to store unordered sets of stack elements at types as well (I can't use always a ``std::set``, since the comparison operators are non-deterministic and it's highly non-trivial to change them to *be* deterministic). Also I don't know if everybody who will ever use this thing will read the comment to stick to it, so I formulated it cautiously :-).